lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp

```diff
@@ -0,0 +1,181 @@
+
+namespace circt {
+namespace firrtl {
+
+/// Deserialize a JSON value into a Target-keyed array of Annotations
+/// (represented as DictionaryAttrs).
+bool fromJSON(llvm::json::Value &value,
+              llvm::StringMap<llvm::SmallVector<DictionaryAttr>> &annotationMap,
+              llvm::json::Path path, MLIRContext *context) {
+
+  /// Examine an Annotation JSON object and return an optional string indicating
+  /// the target associated with this annotation.  Erase the target from the
+  /// JSON object if a target was found.  Automatically convert any legacy Named
+  /// targets to actual Targets.  Note: it is expected that a target may not
+  /// exist, e.g., any subclass of firrtl.annotations.NoTargetAnnotation will
+  /// not have a target.
+  auto findAndEraseTarget =
+      [](llvm::json::Object *object,
+         llvm::json::Path p) -> llvm::Optional<std::string> {
+    // If no "target" field exists, then promote the annotation to a
+    // CircuitTarget annotation by returning a target of "~".
+    auto target = object->get("target");
+    if (!target)
+      return llvm::Optional<std::string>("~");
+
+    // Find the target.
+    auto maybeTarget = object->get("target")->getAsString();
+
+    // If this is a normal Target (not a Named), erase that field in the JSON
+    // object and return that Target.
+    std::string newTarget;
+    if (maybeTarget.getValue()[0] == '~') {
+      newTarget = maybeTarget->str();
+    } else {
+      // This is a legacy target using the firrtl.annotations.Named type.  This
+      // can be trivially canonicalized to a non-legacy target, so we do it with
+      // the following thre mappings:
+      //   1. CircuitName => CircuitTarget, e.g., A -> ~A
+      //   2. ModuleName => ModuleTarget, e.g., A.B -> ~A|B
+      //   3. ComponentName => ReferenceTarget, e.g., A.B.C -> ~A|B>C
+      newTarget = "~";
+      llvm::raw_string_ostream s(newTarget);
+      bool isModule = true;
+      for (auto a : maybeTarget.getValue()) {
+        switch (a) {
+        case '.':
+          if (isModule) {
+            s << "|";
+            isModule = false;
+            break;
+          }
+          s << ">";
+          break;
+        default:
+          s << a;
+        }
+      }
+    }
+
+    // If the target is something that we know we don't support, then error.
+    bool unsupported =
+        std::any_of(newTarget.begin(), newTarget.end(), [](char a) {
+          return a == '/' || a == ':' || a == '>' || a == '.' || a == '[';
+        });
+    if (unsupported) {
+      p.field("target").report(
+          "Unsupported target (not a CircuitTarget or ModuleTarget)");
+      return {};
+    }
+
+    // Remove the target field from the annotation and return the target.
+    object->erase("target");
+    return llvm::Optional<std::string>(newTarget);
+  };
+
+  /// Convert arbitrary JSON to an MLIR Attribute.
+  std::function<Attribute(llvm::json::Value &, llvm::json::Path)>
+      JSONToAttribute =
```

Comment:
nit plz name functions/variables with lower case letter, maybe `convertJSONToAttribute`

Reply:
Changed to use the `convertJSONToAttribute` suggestion.